### ISSUE TITLE
Precompile fails on 32bit linux

### ISSUE THREAD

**Issue by kagalenko-m-b** [JuliaIO/HDF5.jl#789]
Removed ```.julia``` directory and then tried to install and precompile HDF5:
fails complaining about undefined libhdf5
 
```
@v1.5) pkg> add HDF5
 Installing known registries into `~/.julia`
######################################################################## 100,0%
      Added registry `General` to `~/.julia/registries/General`
  Resolving package versions...
  Installed LibCURL_jll ─ v7.70.0+2
  Installed Blosc_jll ─── v1.14.3+1
  Installed HDF5_jll ──── v1.12.0+1
  Installed nghttp2_jll ─ v1.40.0+2
  Installed Zlib_jll ──── v1.2.11+18
  Installed Compat ────── v3.25.0
  Installed HDF5 ──────── v0.14.3
  Installed JLLWrappers ─ v1.1.3
  Installed Requires ──── v1.1.2
  Installed Artifacts ─── v1.3.0
  Installed MbedTLS_jll ─ v2.16.8+1
  Installed Zstd_jll ──── v1.4.5+2
  Installed OpenSSL_jll ─ v1.1.1+6
  Installed Lz4_jll ───── v1.9.2+2
  Installed LibSSH2_jll ─ v1.9.0+3
  Installed Blosc ─────── v0.7.0
Downloading artifact: LibCURL
Downloading artifact: Blosc
Downloading artifact: Zlib
Downloading artifact: nghttp2
Downloading artifact: MbedTLS
Downloading artifact: Zstd
Downloading artifact: LibSSH2
Downloading artifact: OpenSSL
Downloading artifact: Lz4
Updating `~/.julia/environments/v1.5/Project.toml`
  [f67ccb44] + HDF5 v0.14.3
Updating `~/.julia/environments/v1.5/Manifest.toml`
  [56f22d72] + Artifacts v1.3.0
  [a74b3585] + Blosc v0.7.0
  [0b7ba130] + Blosc_jll v1.14.3+1
  [34da2185] + Compat v3.25.0
  [f67ccb44] + HDF5 v0.14.3
  [0234f1f7] + HDF5_jll v1.12.0+1
  [692b3bcd] + JLLWrappers v1.1.3
  [deac9b47] + LibCURL_jll v7.70.0+2
  [29816b5a] + LibSSH2_jll v1.9.0+3
  [5ced341a] + Lz4_jll v1.9.2+2
  [c8ffd9c3] + MbedTLS_jll v2.16.8+1
  [458c3c95] + OpenSSL_jll v1.1.1+6
  [ae029012] + Requires v1.1.2
  [83775a58] + Zlib_jll v1.2.11+18
  [3161d3a3] + Zstd_jll v1.4.5+2
  [8e850ede] + nghttp2_jll v1.40.0+2
  [2a0f44e3] + Base64
  [ade2ca70] + Dates
  [8bb1440f] + DelimitedFiles
  [8ba89e20] + Distributed
  [b77e0a4c] + InteractiveUtils
  [76f85450] + LibGit2
  [8f399da3] + Libdl
  [37e2e46d] + LinearAlgebra
  [56ddb016] + Logging
  [d6f4376e] + Markdown
  [a63ad114] + Mmap
  [44cfe95a] + Pkg
  [de0858da] + Printf
  [3fa0cd96] + REPL
  [9a3f8284] + Random
  [ea8e919c] + SHA
  [9e88b42a] + Serialization
  [1a1011a3] + SharedArrays
  [6462fe0b] + Sockets
  [2f01184e] + SparseArrays
  [10745b16] + Statistics
  [8dfed614] + Test
  [cf7118a7] + UUIDs
  [4ec0a83e] + Unicode
   Building HDF5 → `~/.julia/packages/HDF5/d0V7K/deps/build.log`

(@v1.5) pkg> precompile
Precompiling project...
[ Info: Precompiling HDF5 [f67ccb44-e63f-5c2f-98bd-6dc0ccc4ba2f]
ERROR: LoadError: LoadError: UndefVarError: libhdf5 not defined
Stacktrace:
 [1] top-level scope at /home/kmb/.julia/packages/HDF5/d0V7K/src/api_types.jl:111
 [2] include(::Function, ::Module, ::String) at ./Base.jl:380
 [3] include at ./Base.jl:368 [inlined]
 [4] include(::String) at /home/kmb/.julia/packages/HDF5/d0V7K/src/HDF5.jl:1
 [5] top-level scope at /home/kmb/.julia/packages/HDF5/d0V7K/src/HDF5.jl:51
 [6] include(::Function, ::Module, ::String) at ./Base.jl:380
 [7] include(::Module, ::String) at ./Base.jl:368
 [8] top-level scope at none:2
 [9] eval at ./boot.jl:331 [inlined]
 [10] eval(::Expr) at ./client.jl:467
 [11] top-level scope at ./none:3
in expression starting at /home/kmb/.julia/packages/HDF5/d0V7K/src/api_types.jl:111
in expression starting at /home/kmb/.julia/packages/HDF5/d0V7K/src/HDF5.jl:51

julia> versioninfo()
Julia Version 1.5.3
Commit 788b2c77c1 (2020-11-09 13:37 UTC)
Platform Info:
  OS: Linux (i686-pc-linux-gnu)
  CPU: Intel(R) Pentium(R) 4 CPU 3.00GHz
  WORD_SIZE: 32
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, prescott)
```

**Comment by musm** [comment]
Yeah the issue with HDF5_jll v1.12.0 is we had to drop 32-bit support on linux (conda-forge stopped shipping them), and the Pkg mechanism doesn't (by design) allow you to restrict upgrades based on platform.

We should think about a way to more elegantly handle this. For now, your best bet is to pin HDF5v_jll to `v1.10.5` or use system libraries as instructed in  https://github.com/JuliaIO/HDF5.jl#installation

**Comment by kagalenko-m-b** [comment]
> Yeah the issue with HDF5_jll v1.12.0 is we had to drop 32-bit support on linux (conda-forge stopped shipping them)

I don't understand what conda-forge's role is in that. HDF5_jll is automatically build based on Yggdrasil build_tarballs.jl script. It doesn't [require](https://github.com/JuliaPackaging/Yggdrasil/blob/39838bbfe874d9bcb0132087a4e29f873fb11aaa/H/HDF5/build_tarballs.jl#L93) building on all supported platforms. If that line is changed to
```
platforms = supported_platforms() 
```
that should solve the issue.

**Comment by musm** [comment]
HDF5_jll does not directly compile the libhdf5 library, since libhdf5 does not support cross-compilation so we must resort to other options. What HDF5_jll does is use the uploaded binaries from conda-forge and unpacks them.  

Passing `platforms = supported_platforms() ` would fail for your platform. The issue is that we don't have a way to prevent HDF5 upgrading to  `HDF5_JLL` to `v1.12.0`  on platforms that are not supported  by `v1.12.0`, since the resolver doesn't have a way to prevent upgrading a package based on the platform restriction, by design. This is a unique situation since for most binaries it's not an issue since cross-compilation is feasible.

**Comment by kagalenko-m-b** [comment]
> HDF5_jll does not directly compile the libhdf5 library, since libhdf5 does not support cross-compilation so we must resort to other options.

Have you looked at using BinaryBuilder instead of conda? It compiles for all platforms in virtual environment,
co no cross-compilation needed.

**Comment by kagalenko-m-b** [comment]
I guess I don't understand why are you using conda, current version of BinaryBuilder can compile binary
artifacts in virtual environment for all needed platforms.

**Comment by musm** [comment]
We can't because libhdf5 does not support cross-compilation, and this is how BinaryBuilder compiles the builds for  various platforms

**Comment by musm** [comment]
Just to be clear if it wasn't. We are using BB but using it distribute pre-built binaries in a stable manner.

**Comment by kagalenko-m-b** [comment]
BinaryBuilder compiles artifacts in virtualized environment(s), if a library builds under Linux, then BB can compile it. So I still don't understand why you speak of libhdf5's support for cross-compilation. None is required.

**Comment by musm** [comment]
You are welcome to look at the BinaryBuilder script I co-authored if you don't believe me https://github.com/JuliaPackaging/Yggdrasil/blob/master/H/HDF5/build_tarballs.jl 

We still need cross-compilation because BinaryBuilder uses a Linux host environment

**Comment by jmert** [comment]
I haven't yet dove into figuring out BinaryBuilder/Yggdrasil, so correct me where I'm assuming something wrong: I'm guessing it's cross-compiling for Windows and/or OSX which is the problem (just since that's a much larger change than Linux 64 vs 32bit) — would it be possible/prudent to cross compile Linux and still rely on the pre-compiled binaries from Conda for Windows and Mac?

**Comment by musm** [comment]
Turns out that it's not only an issue for Windows/OSX, which might be expected, but also Linux on non-host architectures. 
E.g. cross-compiling for Linux on arm doesn't work, since the build process interrogates the system for variables that are target dependent.  

There have been some efforts to reverse engineer these variables (i.e. I created a repo to collect some of these so that steven-varga could reverse engineer the process https://github.com/musm/hdf5-build-output) I'm not sure on the status of that effort or even if it will be the easiest approach. stevengjohson went a different route https://github.com/stevengj/hdf5/commits/cross (which looks more straightforward)/
